pkg/store/bucket.go

```diff
@@ -29,101 +29,132 @@ import (
 
-	// Class B operation.
-	gcsOperationObjectGet = "object.get"
-)
+	// Get returns a new reader against the object with the given name.
+	Get(ctx context.Context, name string) (io.ReadCloser, error)
+
+	// GerRange returns a new reader against the object that reads len bytes
+	// starting at off.
+	GetRange(ctx context.Context, name string, off, len int64) (io.ReadCloser, error)
+}
+
+// BucketWithMetrics takes a bucket and registers metrics with the given registry for
+// operations run against the bucket.
+func BucketWithMetrics(name string, b Bucket, r prometheus.Registerer) Bucket {
+	bkt := &metricBucket{
+		Bucket: b,
+		ops: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_store_bucket_operations_total",
+			Help:        "Total number of operations that were executed against a bucket.",
+			ConstLabels: prometheus.Labels{"bucket": name},
+		}, []string{"operation"}),
+	}
+	if r != nil {
+		r.MustRegister(bkt.ops)
+	}
+	return bkt
+}
+
+type metricBucket struct {
+	Bucket
+	ops *prometheus.CounterVec
+}
 
-// GCSStore implements the store API backed by a GCS bucket. It loads all index
+func (b *metricBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	b.ops.WithLabelValues("get").Inc()
+	return b.Bucket.Get(ctx, name)
+}
+
+func (b *metricBucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	b.ops.WithLabelValues("get_range").Inc()
+	return b.Bucket.GetRange(ctx, name, off, length)
+}
+
+// BucketStore implements the store API backed by a Bucket bucket. It loads all index
 // files to local disk.
-type GCSStore struct {
+type BucketStore struct {
 	logger  log.Logger
-	metrics *gcsStoreMetrics
-	bucket  *storage.BucketHandle
+	metrics *bucketStoreMetrics
+	bucket  Bucket
 	dir     string
 
 	mtx                sync.RWMutex
-	blocks             map[ulid.ULID]*gcsBlock
+	blocks             map[ulid.ULID]*bucketBlock
 	gossipTimestampsFn func(mint int64, maxt int64)
 
 	oldestBlockMinTime   int64
 	youngestBlockMaxTime int64
 }
 
-var _ storepb.StoreServer = (*GCSStore)(nil)
+var _ storepb.StoreServer = (*BucketStore)(nil)
 
-type gcsStoreMetrics struct {
+type bucketStoreMetrics struct {
 	blockDownloads           prometheus.Counter
 	blockDownloadsFailed     prometheus.Counter
 	seriesPrepareDuration    prometheus.Histogram
 	seriesPreloadDuration    prometheus.Histogram
 	seriesPreloadAllDuration prometheus.Histogram
 	seriesMergeDuration      prometheus.Histogram
-	gcsOperations            *prometheus.CounterVec
 }
 
-func newGCSStoreMetrics(reg *prometheus.Registry, s *GCSStore) *gcsStoreMetrics {
-	var m gcsStoreMetrics
+func newBucketStoreMetrics(reg *prometheus.Registry, s *BucketStore) *bucketStoreMetrics {
+	var m bucketStoreMetrics
 
 	m.blockDownloads = prometheus.NewCounter(prometheus.CounterOpts{
-		Name: "thanos_gcs_store_block_downloads_total",
+		Name: "thanos_bucket_store_block_downloads_total",
```

Comment:
you have also `thanos_store_bucket_operations_total`. Lets agree either on `thanos_store_bucket_...` or `thanos_bucket_store...` I don't mind which one

Reply:
I believe some of the metrics aren't even used anymore. Let's revisit in a subsequent PR.